File: dataforge-fx/src/main/java/hep/dataforge/fx/configuration/ConfigFXTreeItem.java
package hep.dataforge.fx.configuration;

import hep.dataforge.description.NodeDescriptor;
import hep.dataforge.meta.Configuration;
import javafx.event.Event;
import javafx.scene.Node;
import javafx.scene.control.TreeItem;

import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * Created by darksnake on 01-May-17.
 */
public class ConfigFXTreeItem extends TreeItem<ConfigFX> {
    /**
     * The tag not to display node or value in configurator
     */
    public static final String NO_CONFIGURATOR_TAG = "nocfg";


    public ConfigFXTreeItem(ConfigFX value) {
        super(value);
        fillChildren();
        value.addObserver((o, arg) -> invalidate());
    }

    public ConfigFXTreeItem(ConfigFX value, Node graphic) {
        super(value, graphic);
        fillChildren();
        value.addObserver((o, arg) -> invalidate());
    }

    private Function<Configuration, NodeDescriptor> descriptorProvider;

    private void fillChildren() {
        getChildren().setAll(
                ConfigFXTreeItem.this.getValue().getChildren().stream()
                        .filter(cfg -> toShow(cfg))
                        .map(ConfigFXTreeItem::new)
                        .collect(Collectors.toList())
        );
    }

    @Override
    public boolean isLeaf() {
        return getValue() instanceof ConfigFXValue;
    }


    private boolean toShow(ConfigFX cfg) {
        if (cfg instanceof ConfigFXNode) {
            return !((ConfigFXNode) cfg).getDescriptor().map(nd -> nd.tags().contains(NO_CONFIGURATOR_TAG)).orElse(false);
        } else if (cfg instanceof ConfigFXValue) {
            return !((ConfigFXValue) cfg).getDescriptor().map(vd -> vd.tags().contains(NO_CONFIGURATOR_TAG)).orElse(false);
        } else {
            return true;
        }
    }


    /**
     * Is this branch a root
     *
     * @return
     */
    public boolean isRoot() {
        return this.getParent() == null;
    }

    public void invalidate(){
        fillChildren();
        TreeModificationEvent<ConfigFX> event = new TreeModificationEvent<>(TreeItem.valueChangedEvent(), this);
        Event.fireEvent(this, event);
    }

}

File: kodex/kodex-gui/src/main/kotlin/hep/dataforge/fx/configuration/ConfigFXTreeItem.kt
package hep.dataforge.fx.configuration

import hep.dataforge.description.NodeDescriptor
import hep.dataforge.meta.Configuration
import javafx.event.Event
import javafx.scene.Node
import javafx.scene.control.TreeItem

import java.util.function.Function
import java.util.stream.Collectors

/**
 * Created by darksnake on 01-May-17.
 */
class ConfigFXTreeItem(value: ConfigFX, graphic: Node? = null) : TreeItem<ConfigFX>(value, graphic) {

    private val descriptorProvider: Function<Configuration, NodeDescriptor>? = null


    /**
     * Is this branch a root
     *
     * @return
     */
    val isRoot: Boolean
        get() = this.parent == null

    init {
        fillChildren()
        value.addObserver { o, arg -> invalidate() }
    }


    private fun fillChildren() {
        children.setAll(
                value.getChildren().stream()
                        .filter { cfg -> toShow(cfg) }
                        .map { ConfigFXTreeItem(it) }
                        .collect(Collectors.toList())
        )
    }

    override fun isLeaf(): Boolean {
        return value is ConfigFXValue
    }


    private fun toShow(cfg: ConfigFX): Boolean {
        return when (cfg) {
            is ConfigFXNode -> cfg.descriptor?.tags()?.contains(NO_CONFIGURATOR_TAG) ?: false
            is ConfigFXValue -> cfg.descriptor?.tags()?.contains(NO_CONFIGURATOR_TAG) ?: false
            else -> true
        }
    }

    private fun invalidate() {
        fillChildren()
        val event = TreeItem.TreeModificationEvent(TreeItem.valueChangedEvent<Any>(), this)
        Event.fireEvent(this, event)
    }

    companion object {
        /**
         * The tag not to display node or value in configurator
         */
        val NO_CONFIGURATOR_TAG = "nocfg"
    }

}
